weld/jpa/src/main/java/org/jboss/as/weld/services/bootstrap/WeldJpaInjectionServices.java

```diff
@@ -128,4 +144,143 @@ public EntityManagerResourceReferenceFactory(String scopedPuName, EntityManagerF
+                                    entityManagerFactory = persistenceUnitService.getEntityManagerFactory();
+                                    return null;
+                                }
+                            };
+                    WildFlySecurityManager.doChecked(privilegedAction, accessControlContext);
+                    return entityManagerFactory;
+                }
+
+                @Override
+                public void release() {
+                }
+            };
+        }
+    }
+
+    private static class LazyEntityManager<T> implements ResourceReferenceFactory<EntityManager> {
+        private final PersistenceUnitServiceImpl persistenceUnitService;
+        private final ServiceController<?> serviceController;
+        private final String scopedPuName;
+        private final TransactionSynchronizationRegistry transactionSynchronizationRegistry;
+        private final TransactionManager transactionManager;
+        private final PersistenceContext context;
+
+        public LazyEntityManager(PersistenceUnitServiceImpl persistenceUnitService,
+                                 ServiceController<?> serviceController,
+                                 String scopedPuName,
+                                 PersistenceContext context,
+                                 TransactionSynchronizationRegistry transactionSynchronizationRegistry,
+                                 TransactionManager transactionManager) {
+            this.persistenceUnitService = persistenceUnitService;
+            this.serviceController = serviceController;
+            this.scopedPuName = scopedPuName;
+            this.context = context;
+            this.transactionSynchronizationRegistry = transactionSynchronizationRegistry;
+            this.transactionManager = transactionManager;
+        }
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        @Override
+        public ResourceReference<EntityManager> createResource() {
+            serviceController.addListener(
+                    new LifecycleListener() {
+
+                        @Override
+                        public void handleEvent(final ServiceController<?> controller, final LifecycleEvent event) {
+                            if (event == LifecycleEvent.UP) {
+                                latch.countDown();
+                                controller.removeListener(this);
+                            }
+                        }
+                    }
+            );
+            try {
+                latch.await();
+            } catch (InterruptedException e) {
+                // Thread was interrupted, which we will preserve in case a higher level operation needs to see it.
+                Thread.currentThread().interrupt();
+                // rather than just returning the current EntityManagerFactory (might be null or not null),
+                // fail with a runtime exception.
+                throw new RuntimeException(e);
```

Comment:
Do others agree with setting the thread as interrupted for the higher level to possible detect and then failing the current thread as well by throwing runtime exception?

Reply:
Restoring the interrupt state ensures that any higher level check of thread.isInterrupted() will return true.

Since the latch.await() call was interrupted by a different thread for a good reason, it seems that the best outcome is to fail by throwing an exception.

Reply:
+1. Back in the day I learned that it was best practice to restore the interrupted state.